### PR TITLE
Day 15 Ex 01 Answer Question

### PR DESCRIPTION
**Pull request by at-thienngnuyen**
#### Related issues (Optional)
Task #70
#### Description (Required)
- [x] Ui Test
- [x] get data from JSON file
- [x] create DialogFragment
- [x] random question and answerest
#### Reference (Optional)

### CODE REVIEW

File: app/src/main/java/vn/asiantech/internship/ui/questions/QuestionMainFragment.java [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+import java.util.Random;
+
+import vn.asiantech.internship.R;
+import vn.asiantech.internship.models.Question;
+
+/**
+ * QuestionMainFragment of Question
+ * Created by Thanh Thiem
+ */
+public class QuestionMainFragment extends Fragment implements View.OnClickListener{
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/ui/questions/QuestionMainFragment.java]: ERROR: '{' is not preceded with whitespace.

File: app/src/main/java/vn/asiantech/internship/ui/questions/QuestionShowFragment.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
+                if (isChecked) {
+                    sendToMain(mRbD.getText().toString(), mCurrentPosition);
+                }
+            }
+        });
+        return v;
+    }
+
+    private void sendToMain(String s, int mCurrentPosition) {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/ui/questions/QuestionShowFragment.java]: ERROR: 'mCurrentPosition' hides a field.

File: app/src/main/java/vn/asiantech/internship/models/Question.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
+package vn.asiantech.internship.models;
+
+import android.os.Parcel;
+import android.os.Parcelable;
+
+/**
+ * Models Question Created by Thanh Thien on 6/23/17.
+ */
+public class Question implements Parcelable {
+    private String question;
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/models/Question.java]: WARNING: [AvoidFieldNameMatchingTypeName] It is somewhat confusing to have a field name matching the declaring class name
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/naming.html#AvoidFieldNameMatchingTypeName

File: app/src/main/java/vn/asiantech/internship/ui/questions/QuestionMainFragment.java [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+                String answerB = object.getString("answer_b");
+                String answerC = object.getString("answer_c");
+                String answerD = object.getString("answer_d");
+                String answerCorrect = object.getString("answer_right");
+                Question question = new Question(questionString, answerA, answerB, answerC, answerD, answerCorrect);
+                questions.add(question);
+            }
+
+        } catch (JSONException e) {
+            e.printStackTrace();
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/ui/questions/QuestionMainFragment.java]: WARNING: [AvoidPrintStackTrace] Avoid printStackTrace(); use a logger call instead.
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/logging-java.html#AvoidPrintStackTrace

File: app/src/main/java/vn/asiantech/internship/ui/questions/QuestionMainFragment.java [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+        String json;
+        try {
+            InputStream is = getActivity().getAssets().open("questions.json");
+            int size = is.available();
+            byte[] buffer = new byte[size];
+            is.read(buffer);
+            is.close();
+            json = new String(buffer, "UTF-8");
+        } catch (IOException ex) {
+            ex.printStackTrace();
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/ui/questions/QuestionMainFragment.java]: WARNING: [AvoidPrintStackTrace] Avoid printStackTrace(); use a logger call instead.
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/logging-java.html#AvoidPrintStackTrace

File: app/src/main/res/layout/fragment_content_select.xml [hunk 6]
```diff
@@ -46,11 +46,11 @@
         </LinearLayout>
 
         <Button
-            android:id="@+id/btnEx08"
+            android:id="@+id/btnEx15"
```
Comment (at-dev-support-01) [inline app/src/main/res/layout/fragment_content_select.xml]: WARNING: [InconsistentLayout] The id "btnEx15" in layout "fragment_content_select" is missing from the following layout configurations: layout-v21 (present in layout)
 This check ensures that a layout resource which is defined in multiple resource folders, specifies the same set of widgets.

This finds cases where you have accidentally forgotten to add a widget to all variations of the layout, which could result in a runtime crash for some resource configurations when a `findViewById()` fails.

There *are* cases where this is intentional. For example, you may have a dedicated large tablet layout which adds some extra widgets that are not present in the phone version of the layout. As long as the code accessing the layout resource is careful to handle this properly, it is valid. In that case, you can suppress this lint check for the given extra or missing views, or the whole layout

File: app/src/main/res/layout/fragment_question_result.xml [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        android:id="@+id/toolbar"
+        android:layout_width="match_parent"
+        android:layout_height="50dp"
+        android:background="@color/colorPrimary">
+
+        <RelativeLayout
+            android:layout_width="match_parent"
+            android:layout_height="match_parent">
+
+            <TextView
```
Comment (at-dev-support-01) [inline app/src/main/res/layout/fragment_question_result.xml]: WARNING: [RelativeOverlap] `@id/tvHead` can overlap `@id/imgTryAgain` if @string/result grows due to localized text expansion
 If relative layout has text or button items aligned to left and right sides they can overlap each other due to localized text expansion unless they have mutual constraints like `toEndOf`/`toStartOf`.

File: app/src/main/res/layout/item_question_result.xml [hunk 8]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:id="@+id/rlParent"
+    android:layout_width="match_parent"
+    android:layout_height="45dp">
+
+    <TextView
```
Comment (at-dev-support-01) [inline app/src/main/res/layout/item_question_result.xml]: WARNING: [RelativeOverlap] `@id/tvNumQuestion` can overlap `@id/imgCheck` if @id/tvNumQuestion grows due to localized text expansion
 If relative layout has text or button items aligned to left and right sides they can overlap each other due to localized text expansion unless they have mutual constraints like `toEndOf`/`toStartOf`.

File: app/src/main/java/vn/asiantech/internship/ui/questions/QuestionMainFragment.java [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+import java.util.Random;
+
+import vn.asiantech.internship.R;
+import vn.asiantech.internship.models.Question;
+
+/**
+ * QuestionMainFragment of Question
+ * Created by Thanh Thiem
+ */
+public class QuestionMainFragment extends Fragment implements View.OnClickListener{
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/ui/questions/QuestionMainFragment.java]: ERROR: '{' is not preceded with whitespace.

File: app/src/main/java/vn/asiantech/internship/models/Question.java [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/models/Question.java]: WARNING: [AvoidFieldNameMatchingTypeName] It is somewhat confusing to have a field name matching the declaring class name
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/naming.html#AvoidFieldNameMatchingTypeName

File: app/src/main/res/layout/item_question_result.xml (hunk 8 above)
Comment (at-dev-support-01) [inline app/src/main/res/layout/item_question_result.xml]: WARNING: [RelativeOverlap] `@id/tvNumQuestion` can overlap `@id/imgCheck` if @id/tvNumQuestion grows due to localized text expansion
 If relative layout has text or button items aligned to left and right sides they can overlap each other due to localized text expansion unless they have mutual constraints like `toEndOf`/`toStartOf`.

File: app/src/main/java/vn/asiantech/internship/ui/questions/QuestionMainFragment.java (hunk 9 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/ui/questions/QuestionMainFragment.java]: ERROR: '{' is not preceded with whitespace.

File: app/src/main/java/vn/asiantech/internship/models/Question.java (hunk 10 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/models/Question.java]: WARNING: [AvoidFieldNameMatchingTypeName] It is somewhat confusing to have a field name matching the declaring class name
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/naming.html#AvoidFieldNameMatchingTypeName

**Comment by at-thienngnuyen** [comment]
i fixed . please check, thank you :)

**Comment by at-tamnguyen2** [comment]
@at-thienngnuyen 
LGTM
Thank you

**Comment by at-tienhoang** [comment]
LGTM! thank you
LGTM

**Pull request merged**